Multilib support

stilor opened the pull request:
Lots of changes; see individual commits. Added support for multilib for uClibc and musl; fixed glibc (based on @mingwandroid's branch). Multiple quirks for various x86/mips/powerpc/sh4/sparc multilib quirks. ARM added as a separate commit because it only has all-or-nothing multilib configuration, and the "all" variant includes architectures which are not fully supported by either libc variant (armv4/thumb is not supported at all, and armv8 is not supported by uClibc).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/crosstool-ng/crosstool-ng/383)
<!-- Reviewable:end -->

stilor commented:
Forgot to mention: it's based on the #373 PR.

mingwandroid commented:
This is a huge amount of work @stilor, good to see.

bhundven commented:
@mingwandroid So is your comment a review?

bhundven commented:
@stilor I'll start going through and reviewing these tomorrow.

mingwandroid commented:
It's not a review. I'll attempt to test it this weekend.

bhundven commented:
@mingwandroid Cool. I'm reviewing this tonight and running a few tests.

mingwandroid commented:
I should have said, I did a partial review, eyeballing each change, so it wasn't completely vacuous.

stilor commented:
Rebased onto master.

bhundven commented on scripts/build/arch/alpha.sh:
```diff
@@ -4,3 +4,16 @@ CT_DoArchTupleValues () {
     # The architecture part of the tuple:
```
My comment with this patch is specifically with the commit message:
I would change:

  It doesn't seem right having architecture dependent code in something generic like the (e)glibc-common script.

to:

  It doesn't seem right having architecture dependent code a specific libc implementation script.

Sine the (e)glibc-common script no longer exists.

stilor replied:
At the time @mingwandroid committed this, eglibc was still in the tree, I think. His branch has been sitting for quite some time. I'll change the commit message, but I think it's nitpicking :)

bhundven replied:
It helps for bisecting and general "trying to find out where things changed". If I see (e)glibc-common I start thinking of older code that isn't relative anymore. Yes, it's a bit of nitpicking, but theres a method to the madness.

bhundven commented on scripts/build/libc/glibc.sh:
```diff
@@ -72,6 +72,7 @@ do_libc_backend() {
     local multilib
     local multi_dir
     local multi_flags
+    local target
```
should this be ```libc_target```?

stilor replied:
No, *this* function sets ${target}, ${libc_target} is used in the called function (do_libc_backend_once).

bhundven commented on scripts/build/libc/glibc.sh:
```diff
@@ -72,6 +72,7 @@ do_libc_backend() {
+    local multi_last
     local target
     local extra_dir
     local target
```
Again, two ```local target```?

stilor replied:
This code is gone in favor of a common iterator anyway.

bhundven commented on scripts/build/libc/glibc.sh:
```diff
@@ -167,7 +175,8 @@ do_libc_backend() {
                              libc_startfiles="${libc_startfiles}"   \
                              libc_full="${libc_full}"               \
-                             libc_target="${target}"
+                             libc_target="${target}"                \
```
libc_target?

stilor replied:
No, it sets ${target} above but passes it down as ${libc_target}.
It's mingwandroid's original change, I picked it up as is (mostly, so long as that worked :)) This code was later changed into a common iterator anyway.

bhundven commented on scripts/crosstool-NG.sh.in:
```diff
@@ -292,7 +292,7 @@ if [ -z "${CT_RESTART}" ]; then
     esac
 
-    # Arrange paths depending on wether we use sysroot or not.
+    # Arrange paths depending on whether we use sysroot or not.
```
No rams here :D

stilor commented:
Let's re-run the CI build because of a changed commit message :)

bhundven commented:
Well, my review is not complete, and I'd hope that you'd make more then one change per rebase :)

bhundven commented on scripts/build/arch.sh:
```diff
@@ -57,5 +57,12 @@ CT_DoArchUClibcHeaderDir() {
+# Multilib/uClibc: Adjust header installation path for given CFLAGS
+# Usage: CT_DoArchUClibcHeaderDir <path-variable> <cflags>
+CT_DoArchMUSLHeaderDir() {
+    # Only needed if a given architecture may select different uClibc architectures.
```
s/uClibc/musl/

bhundven commented:
Beyond the review I have done today, I would still like to hear feedback from @mingwandroid, as I cannot validate for mingw-w64 support.

stilor commented:
Not sure why github no longer shows the comments after a branch is rebased (rebased this branch to master, again). I addressed s/uClibc/MUSL/ comment.

yann-morin-1998 commented on config/libc/musl.in:
```diff
@@ -1,6 +1,7 @@
 # musl options
 
 ## depends on ! WINDOWS && ! BARE_METAL
+## depends on EXPERIMENTAL
```
An alternate solution would have been to have musl depend on the
archictectures we know it works on (or excluded on the architectures
we know are broken).

Right now, Buildroot considered musl to be no longer experimental,
so it's weird we would do the opposite... But indeed, Buildroot has
a rather strict set of architectures on which musl is supposed to be
supported:
    https://git.buildroot.org/buildroot/tree/toolchain/toolchain-buildroot/Config.in#n89

But as you wrote, when/if someone is that interested in musl, he can
come up with a patch to fix the situation (I'm not much interested
myself...).

Reviewed-by: "Yann E. MORIN" <yann.morin.1998@free.fr>

yann-morin-1998 commented on scripts/build/arch/sparc.sh:
```diff
@@ -2,4 +2,17 @@
+    # sparc64-*-linux configuration: https://patchwork.ozlabs.org/patch/409424/
+    # but that patch has not been integrated (yet). One concern raised about
+    # this patch was that -mcpu=ultrasparc can suboptimally schedule instructions
+    # for newer SPARC CPUs. So, override to -mcpu=ultrasparc and warn the user.
```
Sorry, I'm not entirely sure I grok this explanation. Let me try to
suggest an alternate commit log:

    arch/sparc: better default CPU when targetting Linux

    By default, sparc64-*-linux is configured with -mcpu=v9. However,
    according to https://sourceware.org/ml/libc-alpha/2005-12/msg00027.html:
        "There is no Linux sparc64 port that runs on non-UltraSPARC-I+ ISA
         CPUs."

    v9 is such a "non-UltraSPARC-I+ ISA CPU", so it makes no sense to
    default to v9 when targetting Linux.

    Change the default to ultrasparc, even though it can suboptimally
    schedule instructions for newer SPARC CPUs. See the pending patch:
        https://patchwork.ozlabs.org/patch/409424/

If that's indeed what you meant, feel free to use it. If nopt, please
explain in more details...

Also, the discussion on the proposed patch suggests that ultrasparc3
would be a better default. Why not use it instead?

stilor replied:
Yes, your understanding is correct. As to ultrasparc3, I am not a SPARC expert (i.e., does ultrasparc3 allow for some instructions that ultrasparc does not?) and I don't have the hardware to test either - so I went with the more conservative option.

yann-morin-1998 replied:
OK, if you're not sure about ultrasparc3, let's keep ultrasparc. The
warning is enough to make the user *think* of a better CPU to set if needed.

bhundven replied:
Is it possible to support both? Just create a tristate in ```config/...``` of the sparc multilib that we support/test in samples?

Probably only need one of the samples in CI, as long as someone can run other samples and tell us if theirs is not supported or is broken.

stilor replied:
Support both **what**? The change does prohibit any explicit CPU setting, it just changes the default for `*-*-linux` configs in case the CPU is not explicitly specified.

bhundven replied:
Ah, ok. I'm still reading through this and understanding. That makes sense!

yann-morin-1998 commented on scripts/build/arch/sparc.sh:
```diff
@@ -2,4 +2,17 @@
+    # but that patch has not been integrated (yet). One concern raised about
+    # this patch was that -mcpu=ultrasparc can suboptimally schedule instructions
+    # for newer SPARC CPUs. So, override to -mcpu=ultrasparc and warn the user.
+    if [ "${CT_KERNEL}" = "linux" -a "${CT_ARCH_64}" = "y" -a "${CT_ARCH_CPU}" = "" ]; then
```
My personal taste is not to check against the empty string (`"${foo}" = ""`), but to check emptyness (`-z "${foo}"`).

yann-morin-1998 commented on config/arch/sh.in:
```diff
@@ -5,6 +5,7 @@
 ## select ARCH_USE_MMU
 ## select ARCH_SUPPORTS_BOTH_ENDIAN
 ## select ARCH_DEFAULT_LE
+## select ARCH_REQUIRES_MULTILIB
```
I'm not sure why we need to have `ARCH_REQUIRES_MULTILIB` when its sole
reason to exist is to select `MULTILIB`. Why not select `MULTILIB` directly?

Well, I can see that is is more meaning full to say "arch requires multilib" rather
than just "force multilib" on; it gives the reason. But still...

Anyway, that's not this patch that introduces it, so:

Reviewed-by: "Yann E. MORIN" <yann.morin.1998@free.fr>

bhundven replied:
Reviewed-by: "Bryan Hundven" <bryanhundven@gmail.com>

yann-morin-1998 commented:
I've had time to review only the very first patches tonight. Will continue tomorrow.
Ping me on IRC when you see me to remind me... ;-)

bhundven commented:
Thanks, Yann!

yann-morin-1998 commented on scripts/build/libc/glibc.sh:
```diff
@@ -288,12 +288,12 @@ do_libc_backend_once() {
-            extra_cc_args="${extra_cc_args} ${endian_extra}"
             ;;
         "") extra_cc_args="${extra_cc_args} ${CT_ARCH_ENDIAN_OPT}"
             ;;
```
Reviewed-by: "Yann E. MORIN" <yann.morin.1998@free.fr>

yann-morin-1998 commented on scripts/build/cc/100-gcc.sh:
```diff
@@ -403,10 +403,12 @@ do_gcc_core_backend() {
     fi
 
-    # Some versions of gcc have a deffective --enable-multilib.
+    # Some versions of gcc have a defective --enable-multilib.
```
I think *deficient* would be even more accurate.
But I'm not the native english speaker... ;-)

stilor replied:
English is my 2nd language too, but:
deficient: not having enough of a specified quality or ingredient
defective: imperfect or faulty
Seems like defective is more appropriate (unless we want to state that some versions of GCC don't have enough multilibs) :)

yann-morin-1998 replied:
OK! :-)

yann-morin-1998 commented on scripts/build/cc/100-gcc.sh:
```diff
@@ -403,10 +403,12 @@ do_gcc_core_backend() {
     if [ "${CT_MULTILIB}" != "y" ]; then
         extra_config+=("--disable-multilib")
+    else
+        extra_config+=("--enable-multiarch")
```
Please explain the relations between --enable/disable-multilib and
--enable-disable-multiarch.

Why would we have to --enable-multiarch when CT_MULTILIB = y ?

Do we want to consider that multilib === multiarch from the point of
view of crosstool-NG ? If so, we should state it explicitly, both here
in the code with a big fat comment, *and* in the help text for
CT_MULTILIB in config/target.in

Otherwise, do we need to explicitly --disable-multiarch along with
--disable-multilib (conv. with --enable)> Like so:

    if [ "${CT_MULTILIB}" != "y" ]; then
        extra_config+=("--disable-multilib" "--disable-multiarch")
    else
        extra_config+=("--enable-multilib" "--enable-multiarch")
    fi

Ditto below, of course.

stilor replied:
Mutlilib makes GCC aware of different library locations/sysroots for different sets of flags. Multiarch adds to that an ability to report the configuration tuple corresponding to these flags. So, multiarch only makes sense if you have multilib. The --enable-multiarch option is actually a tri-state, it can be enabled/disabled/left as auto. The default is 'auto', and the 'auto' logic enables it if it finds multiple subdirectories in ${libdir} with C runtime objects (crt1.o, etc). So, in pass-2/final the multiarch was enabled already if multilib was enabled. The problem was with core pass-1: we don't have the CRT installed yet, so multiarch support was not detected. Effectively, this change applies the status quo configuration in pass-2/final to pass-1 as well.

And, as the comment above this conditional states, --enable-multilib is the default, so we don't need to explicitly pass that.

yann-morin-1998 replied:
OK, so why do we have to --enable-multiarch when what we *do* are about
is multilib ?

Again, I fail to understand why --enable-multiarch is the solution to
enabling multilib...

I can see that we may want to add support for multiarch, but that's not
what this series is about, is it?

yann-morin-1998 commented on config/cc/gcc.in.2:
```diff
@@ -37,6 +37,15 @@ config CC_GCC_EXTRA_CONFIG_ARRAY
+      Architecture-specific option of expanding or restricting the list of
+      the multilib variants to be built. Refer to GCC installation manual
+      for the format of this option for a particular architecture.
+
```
What if the list is empty? Waht about adding:

    Leave empty to use the default list for this architecture.

... or something along those lines...

yann-morin-1998 commented on scripts/build/arch/alpha.sh:
```diff
@@ -4,3 +4,16 @@ CT_DoArchTupleValues () {
+    local multi_flags="${1}"
+    local target="${2}"
+
+    :;
```
Absolutely not needed.

yann-morin-1998 commented on scripts/build/arch/arm.sh:
```diff
@@ -39,3 +39,16 @@ CT_DoArchTupleValues() {
+    local multi_flags="${1}"
+    local target="${2}"
+
+    :;
```
Absolutely not needed (ditto for the following commits, of course).

stilor replied:
Agreed, but this was the change in @mingwandroid's branch which I used as a starting point. I only fixed them where they were broken, didn't do any cosmetic changes at this point as I was pretty sure they'll end up heavily modified anyway. In one of the later commits, these functions are dropped on the architectures where they do nothing.

yann-morin-1998 commented on scripts/build/arch/powerpc.sh:
```diff
@@ -26,3 +26,44 @@ CT_DoArchTupleValues () {
+    esac
+
+    # return the target
+    echo "${target}"
```
It took me a while to understand that `CT_DoArchMultilibTarget` can be
called multiple times, with different `multi_flags` each time, and that
we would only get one of each kind with each call, i.e. that we can't
have both `-m32` and `-m64` in the same call. Right?

Please make that explicit in the commit log. Otherwise, one has to look
at the following commits to understand how/when it is called.

yann-morin-1998 commented on scripts/build/arch/x86.sh:
```diff
@@ -20,4 +20,49 @@ CT_DoArchTupleValues() {
+            CT_DoLog ERROR "CT_TARGET: ${CT_TARGET}"
+            CT_Abort "Go read: https://wiki.debian.org/Multiarch/Tuples"
+            ;;
+    esac
```
This should be part of a separate commit.

yann-morin-1998 commented on scripts/build/arch/x86.sh:
```diff
@@ -20,4 +20,49 @@ CT_DoArchTupleValues() {
+        case "$m" in
+            -m32)  bit=32; abi=default;;
+            -m64)  bit=64; abi=default;;
+            -mx32) bit=64; abi=x32;;
```
To be consistent across all the `CT_DoArchMultilibTarget` functions,
what about:

    local bit32=false
    local bit64=false
    local abi_default=false
    local abi_x32=false

    case "$m" in
        -m32)  bit32=true; abi_default=true;;
        -m64)  bit64=true; abi_default=true;;
        -mx32) bit64=true; abi_x32=true;;
    esac

And use them as their counterparts are used in other
`CT_DoArchMultilibTarget` functions:

    case "${target}" in
        x86_64-*)      $bit32 && target=${target/#x86_64-/i386-} ;;
        i[34567]86-*)  $bit64 && target=${target/#i[34567]86-/x86_64-} ;;
    esac

and so on...

yann-morin-1998 commented on scripts/functions:
```diff
@@ -1305,6 +1305,24 @@ CT_DoBuildTargetTuple() {
+    local target="$2"
+    local gcc_multiarch
+
+    gcc_multiarch=$("${CT_TARGET}-gcc" -print-multiarch ${multi_flags})
```
Please enclose any shell expansion between double quotes:

    gcc_multiarch="$("${CT_TARGET}-gcc" -print-multiarch ${multi_flags})"

yann-morin-1998 replied:
Oh, and redirect `stderr` to `/dev/null` for good measure (since the
comment seems to imply that gcc may choke on this).

stilor replied:
But we can't quote ${multi_flags}! This variable may contain multiple options, and we *want* them to result in multiple arguments, or GCC won't accept a single argument like "-m64 -mel" :)

stilor replied:
And, gcc won't choke - it just reports an empty string with 0 exit code if multiarch is not supported on that configuration.

yann-morin-1998 replied:
> we can't quote ${multi_flags}

No we can't. What I was suggesting was that we quote the shell expansion,
like so:  `"$(...)"`

yann-morin-1998 replied:
> gcc won't choke - it just reports an empty string with 0 exit code

OK, so say so in the comment, maybe:

    # This function determines the target tuple for a given set of
    # compiler flags, using either GCC's multiarch feature if supported
    # (if not, gcc prints nothing and returns with 0), falling back to
    # calling the architecture-specific functions.

stilor replied:
Re quoting: it is not necessary, since the expression on the right is a single bash token - `$(...)`. Any whitespace generated as a result of expansion is still retained, e.g.:
````
avn@mistral:~/work/ctng/crosstool-ng$ a="a   b "
avn@mistral:~/work/ctng/crosstool-ng$ b=${a}
avn@mistral:~/work/ctng/crosstool-ng$ echo "[$b]"
[a   b ]
````

If you insist, I'll do that - but I think unnecessary nested quoting just reduces readability.

yann-morin-1998 commented on scripts/functions:
```diff
@@ -1305,6 +1305,24 @@ CT_DoBuildTargetTuple() {
 
+# This function determines the target tuple for a given set of compiler
+# flags, using either GCC's multiarch feature (if supported) or using
+# fallback architecture-specific functions.
```
As far as I understand it, if we have multilib support, we now do have
multi-arch support too (because of "Enable multiarch support in pass-1
GCC"), no?

So, gcc will alwaysupport -print-multiarch. Unless it may return an
empty set, but still have multilib variants?

stilor replied:
... except that some multilib target configurations in GCC do not support multiarch, in which case --enable-multiarch is a no-op. mingw target is one such example. If that's the case, GCC will still process --print-multiarch option, but will just report an empty string. Yes, it's convoluted and GCC's behavior makes me cringe, too.

yann-morin-1998 replied:
> GCC's behavior makes me cringe

Is there a club? Like the Fight Club or something? May I join? ;-)

mingwandroid commented:
Using this PR, today I built a mingw-w64 targeted multilib GCC on Linux with the following config:

```
CT_LOCAL_TARBALLS_DIR="${HOME}/src"
CT_EXPERIMENTAL=y
CT_SAVE_TARBALLS=y
CT_LOG_EXTRA=y
CT_ARCH_64=y
CT_ARCH_x86=y
CT_MULTILIB=y
CT_TARGET_VENDOR="w64"
CT_KERNEL_windows=y
CT_BINUTILS_PLUGINS=y
CT_MINGW_DIRECTX=y
CT_MINGW_DDK=y
CT_CC_LANG_CXX=y
CT_CC_GCC_SJLJ_EXCEPTIONS=y
CT_CC_GCC_MULTILIB_LIST="m64,m32"
```

.. and could build some C++ tests with both `-m32` and `-m64` and they ran fine through wine-staging.

yann-morin-1998 commented on scripts/functions:
```diff
@@ -1305,6 +1305,24 @@ CT_DoBuildTargetTuple() {
+# fallback architecture-specific functions.
+CT_DoMultilibTarget() {
+    local multi_flags="$1"
+    local target="$2"
```
In fact, if you pass the target as first argument, you can then pass all
multi flags ar extra args and store them in an array:

    CT_DoMultilibTarget() {
        local target="${1}"; shift
        local -a multi_flags=( "${@}" )
        local gcc_multiarch
    
        gcc_multiarch="$( "${CT_TARGET}-gcc" -print-multiarch "${multi_flags[@]}" )"
        if [ -n "${gcc_multiarch=}" ]; then
            printf "%s\n" "${gcc_multiarch=}"
        fi

        CT_DoArchMultilibTarget "${target}" "${multi_flags[@]}"
    }

(of course, propagate the target <-> multi_flags switch to the previous
commit, that introduces `CT_DoArchMultilibTarget`.)

(and the reason I like to have spaces after `(` and before `)` is for
readability).

yann-morin-1998 commented on scripts/build/libc/glibc.sh:
```diff
@@ -464,22 +464,23 @@ do_libc_backend_once() {
             # it doesn't matter what it contains.  So, treating '/dev/null'
             # as a C source file, we produce a dummy 'libc.so' in one step
-            CT_DoExecLog ALL "${cross_cc}" -nostdlib        \
+            CT_DoExecLog ALL "${cross_cc}" ${extra_flags}   \
```
>  CT_DoExecLog ALL "${cross_cc}" *${extra_flags}*

Should that not be in a separate commit (after all, it looks like a fix
for an existing bug, even in the previous code) ?

yann-morin-1998 replied:
Doh, I forgot to say it previously: maybe `${extra_flags}` should
go last (i.e. just after `-shared`).

stilor commented:
Today is 50 days since I posted this PR (posted on 04/19). So far, 8 out of 35 commits have been reviewed. Extrapolated ETA for review completion is thus ~170 days (50 / 8 * (35 - 8)). Wake me up around Chrstimas, please.

My mistake, I somehow thought this project is still maintained.

stilor commented:
I have updated the branch - rebased and addressed the comments from Yann Morin. I am not going to make any further updates to this PR; either take it as it is, or reject and close it.

yann-morin-1998 commented on scripts/functions:
```diff
@@ -1305,6 +1305,25 @@ CT_DoBuildTargetTuple() {
+    local -a multi_flags=( "$@" )
+    local gcc_multiarch
+
+    gcc_multiarch=$( "${CT_TARGET}-gcc" -print-multiarch "${multi_flags[@]}" )
```
For good measure, I'd have redirected `stderr` to `/dev/null`,
just in case different versions of gcc behaves differently.

This can however be changed later, if that happens to be the case.

yann-morin-1998 commented on scripts/build/libc/glibc.sh:
```diff
@@ -170,6 +170,9 @@ do_libc_backend() {
 
         CT_Popd
+        if [ "${multi_dir}" != "." ]; then
+            CT_EndStep
```
This missing EndStep is actually caused by d28a3dab (glibc.sh: Use --print-multi-os-directory)
which does remove a EndStep and thus causes the imbalance.

yann-morin-1998 replied:
@bhundven : you could probably fold that directly in the corresponding commit
without the need for @stilor to rebase-and-repush, no?

yann-morin-1998 replied:
As stated in a comment for #d8a3da, it might be better to change the existing `CT_DoStep` into a `CT_DoLog` instead.

yann-morin-1998 commented on scripts/build/libc/glibc.sh:
```diff
@@ -102,7 +103,14 @@ do_libc_backend() {
+            # This is the last multilib build or multilib is '.'
+            # (default target, not multilib)
+            multi_last=y
+        fi
```
This is a bit convoluted. If you'd build it with the first multilib, the code
would be so much simpler!

    first=1
    for multilib blabla; do
        blabla all code
        if [ ${first} -eq 1 ]; then
            blabla build manuals...
        fi
        first=0
    done

Et voilà! ;-)

yann-morin-1998 replied:
BTW...

Yes, it would need to be passed donw to the `do_libc_backend` function,
but rather than calling it `multi_last`, just call it `may_build_manuals`.

bhundven replied:
Were these addressed?

yann-morin-1998 commented on scripts/build/libc/glibc.sh:
```diff
@@ -166,32 +167,10 @@ do_libc_backend() {
-                CT_DoExecLog DEBUG rm -rf "${CT_SYSROOT_DIR}/${multi_dir}"
-            fi # libc_mode == final
-
-            CT_EndStep
```
As already discussed on IRC, this `CT_EndStep` should not be removed, to match
a similarly-guarded `CT_DoStep` on line 108.

However, it might be better to change that `CT_DoStep` into just a `CT_DoLog` which
is probably good enough for what we need.

stilor replied:
I'll just restore the `CT_EndStep`; since each multilib variant has several logged lines, it helps to visually separate them by indenting. Timing also helps (especially for configuration with a lot of multilibs (e.g. ARM with 28 variants).

yann-morin-1998 commented:
Except for a CT_DoStep to change into a CT_DoLog, #d28a3da looks OK.

yann-morin-1998 commented:
#e181893 looks good.

yann-morin-1998 commented on scripts/functions:
```diff
@@ -1183,6 +1183,14 @@ CT_DoConfigSub() {
+# Usage: CT_EnvModify VAR VALUE
+CT_EnvModify() {
+    echo "${1}=\"${2}\"" >> "${CT_BUILD_DIR}/env.modify.sh"
+}
```
I think a better solution would have been something like:

    # Store extra environment for later steps
    CT_StoreEnv() {
        printf "%s='%s'\n" "${1}" "${2}" >>"${CT_BUILD_DIR}/extra-env.sh"
    }

    # Load extra environment from previous steps
    CT_LoadEnv() {
        if [ -f "${CT_BUILD_DIR}/extra-env.sh" ]; then
            . "${CT_BUILD_DIR}/extra-env.sh"
        fi
    }

And then you use CT_LoadEnv in the top-level sub-shell:

    scripts/crosstool-NG.sh.in:
    623        if [ ${do_it} -eq 1 ]; then
    624            ( CT_LoadEnv; do_${step} )

By loading the environment in the sub-shell, you won;t pollute the
top-level environment, while still allowing passing environment from
steps to steps.

Note that the environment file is not removed after being parsed; it is
cumulative: steps can add new environment, valid for the next steps.

yann-morin-1998 replied:
After discussing on IRC, the current proposal is better than mine.

Modifying the environment in that way already allows passing options to following steps, but even more, the so-modified environment is automatically saved for the save/restore mechanism.

yann-morin-1998 commented on scripts/crosstool-NG.sh.in:
```diff
@@ -628,6 +628,14 @@ if [ "${CT_ONLY_DOWNLOAD}" != "y" -a "${CT_ONLY_EXTRACT}" != "y" ]; then
+                . "${CT_BUILD_DIR}/env.modify.sh"
+                CT_DoExecLog DEBUG rm -f "${CT_BUILD_DIR}/env.modify.sh"
+
+            fi
```
See comment about `CT_EnvModify()`, below.

yann-morin-1998 commented on scripts/build/cc/100-gcc.sh:
```diff
@@ -966,38 +1096,22 @@ do_gcc_backend() {
+        CT_DoLog EXTRA "Building the GCC manuals"
+        CT_DoExecLog ALL ${make} pdf html
+        CT_DoLog EXTRA "Installing the GCC manuals"
+        CT_DoExecLog ALL ${make} install-{pdf,html}-gcc
```
Why move the manual generation? It causes extra and unncessary churn, as far as I can see...

stilor replied:
Four months after making the change, I don't remember the exact reason for this. I think it was just separating docs build from the actual compiler build & tweaking. I don't think it will cause any churn - it is going to do the same stuff, just in different order. But I'll restore the previous order.

yann-morin-1998 commented on scripts/build/cc/100-gcc.sh:
```diff
@@ -563,32 +714,14 @@ do_gcc_core_backend() {
+        CT_DoLog EXTRA "Building the GCC manuals"
+        CT_DoExecLog ALL ${make} pdf html
+        CT_DoLog EXTRA "Installing the GCC manuals"
+        CT_DoExecLog ALL ${make} install-{pdf,html}-gcc
```
Ditto, why move the manual generation?

yann-morin-1998 commented on scripts/build/libc/glibc.sh:
```diff
@@ -111,6 +111,24 @@ do_libc_backend() {
+        #   <sysroot>/m4a/usr/lib/m4a/
+        #   <sysroot>/m4a/lib/
+        #   <sysroot>/m4a/usr/lib/
+        multi_flags=$( echo "${multilib#*;}" | ${sed} -r -e 's/@/ -/g;' )
```
`multi_flags` is set but not used. It should probably belong to a later commit, no?

patrickdepinguin commented:
@bhundven What is the status of this PR in your opinion? Does it need more review, test, ... ? Do you see remaining issues with the implementation?

I tested an earlier version of it successfully, after determining the correct configuration to pass for my use case. I would prefer to have a crosstool-ng release with these patches than applying them manually.

bhundven commented:
@patrickdepinguin It looks to me like Yann has given his review and @stilor has some work to do. My opinion is that those issues need to be resolved before this PR is merged.

stilor commented:
@bhundven No, Yann is still going through the rest of the commits. I am holding off rebasing until there's more feedback to address (at which point, I'll probably create a second "batch" of commits in a separate PR). I don't want to rebase the branch after every single comment.

stilor commented:
@yann-morin-1998 Are you done with the review, or are you going to continue?

stilor commented:
Re-pushed with Yann's feedback addressed. Yann has reviewed the first 4 commits:
````
82072d0 multilib: Determine which options may pass through.
cc86d80 glibc: Build manuals and locales last
27fed7e crosstool-NG.sh.in: Don't make lots of symlinks to lib folder
098bd01 glibc.sh: Use --print-multi-os-directory
````
Feedback addressed, except for 'why move the manual generation`. I don't remember the exact reason for moving it after 6 months, sorry. I don't think it is worth moving back and re-testing, though - especially since this is going to change in a later commit, `98e556d Support multilib in sh/uClibc`, which is going to implement a common iterator for multilibs.

mingwandroid commented:
AFAIR, @bhundven did the manual moving change. Is there anything I can do to help get this merged?

bhundven commented:
Yea, that might have been me 😊 

@mingwandroid: It would be great to get a test run from this branch with mingw-w64? Beyond that, if you have any comments for this review, not much. I have a few specific use-cases I need to test, but beyond that there are only a few more commits to review.

stilor commented:
I think @mingwandroid did that already - see his comment from 05/19. There have been some cosmetic changes due to review feedback since then, but nothing major.

bhundven commented on scripts/build/arch/arm.sh:
```diff
@@ -39,3 +39,22 @@ CT_DoArchTupleValues() {
+        CT_KconfigDisableOption "CONFIG_ARM_EABI" "${cfg}"
+        CT_KconfigEnableOption "CONFIG_ARM_OABI" "${cfg}"
+    fi
+}
```
We only support back to 0.9.33.2, and it also doesn't have `CONFIG_ARM_OABI`.
Maybe we should remove it?

bhundven commented on scripts/build/arch/mips.sh:
```diff
@@ -14,3 +14,36 @@ CT_DoArchTupleValues() {
+
+    # FIXME: uClibc (!ng) allows to select ISA in the config; should
+    # match from the selected ARCH_ARCH level... For now, delete and
+    # fall back to default.
```
After we merge this, could you open an issue to track fixing this! :)

bhundven commented on scripts/build/arch/sh.sh:
```diff
@@ -44,9 +44,39 @@ CT_DoArchUClibcConfig() {
     CT_KconfigDisableOption "CONFIG_SH4" "${cfg}"
     CT_KconfigDisableOption "CONFIG_SH4A" "${cfg}"
-    case "${CT_ARCH_SH_VARIAN}" in
+    case "${CT_ARCH_SH_VARIANT}" in
```
Good catch!

stilor replied:
... of what was my typo in a preceding commit.

bhundven commented:
See, it's been a while :)

I'm really not seeing any issues except a few on mac, but that will have to be fixed later.
I'm OK with these changes. I'm going to wait for the travis-ci run to finish, and then merge.

Thank you @stilor for your work and patience!

mingwandroid commented:
Woo hoo! Many thanks @stilor for seeing this through when the rest of us got too busy. I cannot begin to tell you how grateful I am.

Now then, how would people feel about me adding CMake for LLVM + Clang?
